tools/scripts/use-node-version

```diff
@@ -0,0 +1,158 @@
+#!/usr/bin/env bash
+
+# Check whether the current node version matches the .nvmrc version.
+#
+# If not, try to use a locally installed node version manager to set the correct
+# version.
+#
+# Note that we're not checking all possible nvmrc values, just ones that match
+# $(node -v) format (semver).
+
+# helpers for message styling
+blue='\033[0;34m'
+red='\033[0;31m'
+dim='\033[2m'
+bold='\033[1m'
+reset='\033[0m'
+
+nvmrc_version=$(cat .nvmrc)
+node_version="no Node at all"
+
+# check if we're using het correct version of node (if any)
+using_correct_node() {
+	# is _any_ node available?
+	if [[ -x "$(command -v node)" ]]; then
+		# If asdf is loaded, but not the correct version of node, it errors when
+		# you try to use `node` with a message about installing the right
+		# version. If we find this, we're going to install it ourselves, so we
+		# suppress the warning with the "2>/dev/null) || true" bit
+		node_version=$(node -v 2>/dev/null) || true
+		node_version=${node_version:1} # remove the "v" in "v1.2.3"
+	fi
+	if [ "$node_version" == "$nvmrc_version" ]; then
+		echo -e "Using Node ${blue}$node_version${reset}"
+		return 0
+	else
+		# this will also be false if no node is installed
+		return 1
+	fi
+}
+
+# try to load the correct version of node using a local node version manager (if
+# there is one)
+load_correct_node() {
+	# try using asdf
+	if asdf_installed; then
```

Comment:
is the priority of this chain determined by an official CSTI recommendation? it'd be nice to be able to allow users to choose which tool they use instead of choosing for them, say if i had both `asdf` and `fnm` installed..

or at the very we could give them the option of cancelling the process before automatically going ahead and installing with a given tool

Reply:
it's based on the recommendations for the new laptops.

if you do have asdf _and_ fnm installed, you're going to be in a bit of a pickle anyway (i'm guessing asdf's shims will win out?) but if you _have_ got both working, what's the value in us asking which one you want to use?

the intention was to be as unobstructive as possible – just get hold of the right version of node in the way it looks like you're set up to things

Reply:
good point around having both installed — hadn't considered that might be quite a rare scenario.

what did you think about having some kind of user confirmation prompt?

Reply:
what would we prompt for?

Reply:
i was thinking, prompt to confirm that they want to continue with the installation — but i just remembered that fnm does that anyway:
<img width="846" alt="Screenshot 2022-08-31 at 18 36 53" src="https://user-images.githubusercontent.com/1771189/187743369-647fecd3-4e16-417e-9881-45451a4e3f61.png">

Reply:
i'm not sure - is there any value in saying no? i woudn't want to change anything global on the user's system (e.g. setting the system default node) but here we're just using a node version manager they've already installed themselves to install a version of node, i think that's ok?

the intention here is to make the fact that they _need_ to install it a non-issue (in a non-invasive way)

Reply:
it seems to be a fairly standard pattern where if program is about to run a command that'll download a new executable to a computer it gives a choice to say no before doing so (like you get with apt/brew/pacman etc)

Reply:
> if program is about to run a command that'll download a new executable to a computer it gives a choice to say no before doing so (like you get with apt/brew/pacman etc)

is this true? almost everything you download from brew or npm etc comes with dependencies that are installed without prompting, because you've elected to install the thing that depends on them.

if you try to run a task that has a dependency, and we download that dependency – with a notification that that's what we're doing and using the tool you've installed to do that – is that different?

Reply:
> because you've elected to install the thing that depends on them

I'd argue that in my experience it's quite rare that just running `yarn` installs a specific version of node on my computer on top of installing the project dependencies. So the extra invocation of `fnm`,`asdf`,etc.... doesn't fall within the remit i gave my `yarn install` when I elected to install my dependencies.